src/lib.rs

```diff
@@ -263,264 +213,322 @@ where
+            BufStates::BothBusy => write!(f, "BothBusy"),
+            BufStates::Dummy => write!(f, "Dummy"),
         }
     }
+}
 
-    // Add new items to the buffer, send full buffers to the
-    // writer thread.
-    fn process(&mut self) -> Result<(), Error> {
-        let mut buf = Vec::with_capacity(self.buffer_size);
+impl<T, H: BufHandler<T>> BufferStateMachine<T, H> {
+
+    /// Move `items` into the buffer & it. If a buffer is filled by this operation
+    /// then the calling thread will be used to process the full buffer. 
+    fn add_items(&self, items: &mut Vec<T>) -> Result<(), Error> {
+        if self.closed.load(std::sync::atomic::Ordering::Relaxed) {
+            panic!("tried to add items to BufHandler after it was closed");
+        }
 
         loop {
-            match self.rx.recv() {
-                Ok(Some(v)) => {
-                    buf.extend(v);
-                    if buf.len() + self.chunk_size > self.buffer_size {
-                        buf = self.send_buf(buf, false)?;
+            
+            // get mutex on buffers 
+            let mut buffer_state = self.buffer_state.lock().unwrap();
+            let mut current_state = Dummy;
+            std::mem::swap(buffer_state.deref_mut(), &mut current_state);
+
+            // determine new state and any follow-up work
+            let (mut new_state, outcome) = match current_state {
+                FillAndWait(mut f, w) => {
+                    f.extend(items.drain(..));
+                    if f.len() + self.sender_buffer_size >= f.capacity() {
+                        (FillAndBusy(w), Process(f))
+                    } else {
+                        (FillAndWait(f, w), Done)
                     }
                 }
-                Ok(None) => {
-                    self.send_buf(buf, true)?;
-                    break;
+                FillAndBusy(mut f) => {
+                    f.extend(items.drain(..));
+                    if f.len() + self.sender_buffer_size >= f.capacity() {
+                        (BothBusy, Process(f))
+                    } else {
+                        (FillAndBusy(f), Done)
+                    }
                 }
-                Err(_) => break,
+                BothBusy => (BothBusy, Retry),
+                Dummy => unreachable!(),
+            };
+
+            // Fill in the new state.
+            std::mem::swap(buffer_state.deref_mut(), &mut new_state);
+
+            // drop the mutex
+            drop(buffer_state);
+
+            // outcome could be to process a full vec, retry, or be done.
+            match outcome {
+                Process(mut buf_to_process) => {
+                    // process the buffer & return it to the pool.
+                    self.process_buffer(&mut buf_to_process)?;
+                    self.return_buffer(buf_to_process);
+                    break;
+                },
+                Retry => {
+                    thread::yield_now();
```

Comment:
What does this yield do? Isn't the thread going to attempt to acquire the mutex lock in the next iteration right away?

Reply:
Ah got it, it's giving the thread back to the scheduler.

Reply:
I think this communication should be arbitered by a condvar. Retry should be blocking on it while the Process thread issues to it.

Reply:
@nlhepler going to take this as a todo for the future. Things seem to be behaving ok with the current code.